File: app/src/main/java/com/bitflaker/lucidsourcekit/data/FastSineTable.java
package com.bitflaker.lucidsourcekit.data;

import androidx.annotation.NonNull;

public class FastSineTable
{
	private static final FastSineTable instance = new FastSineTable(44100);
	private final float[] sineTable;
	private final int sampleRate;

	private FastSineTable(int sampleRate)
	{
		this.sampleRate = sampleRate;
		sineTable = generateSineTable();
	}

	public static FastSineTable getTable(){
		return instance;
	}

	@NonNull
	private float[] generateSineTable() {
		final float[] sineTable = new float[this.sampleRate];
		float stepSize = (float) (2d * Math.PI / this.sampleRate);
		for (int i = 0; i < this.sampleRate; i++) {
			sineTable[i] = (float) Math.sin(stepSize * ((float)i));
		}
		return sineTable;
	}

	public float sineBySampleRateDeg(long angle)
	{
		return sineTable[(int)(angle % sampleRate)];
	}
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/binauralbeats/FastSineTable.kt
package com.bitflaker.lucidsourcekit.main.binauralbeats

import kotlin.math.sin

class FastSineTable
private constructor(private val sampleRate: Int) {
    companion object {
        val table: FastSineTable = FastSineTable(44100)
    }
    
    private val sineTable = generateSineTable()

    private fun generateSineTable(): FloatArray {
        val sineTable = FloatArray(this.sampleRate)
        val stepSize = (2.0 * Math.PI / this.sampleRate).toFloat()
        for (i in 0..<this.sampleRate) {
            sineTable[i] = sin((stepSize * (i.toFloat())).toDouble()).toFloat()
        }
        return sineTable
    }

    fun sineBySampleRateDeg(angle: Long): Float {
        return sineTable[(angle % sampleRate).toInt()]
    }
}